noir_stdlib/src/lib.nr

```diff
@@ -110,7 +110,43 @@ where
     T: AsPrimitive<Field>,
     Field: AsPrimitive<T>,
 {
-    AsPrimitive::as_(x.as_() * y.as_())
+    // Perform u128 wrapping multiplication by casting x and y to u128.
+    // For u64 and below, the 'high' u64 limbs will be 0 so the u128
+    // multiplication will be simplified to a normal field multiplication
+    // and the cast to the primitive type will perform the wrapping.
+    AsPrimitive::as_(wrapping_mul128(x.as_(), y.as_()))
+}
+
+global two_pow_64: Field = 0x10000000000000000;
+/// Splits a 128 bits number into two 64 bits limbs
+unconstrained fn split64(x: Field) -> (u64, u64) {
+    let lo = x as u64;
+    let hi = ((x - lo as Field) / two_pow_64) as u64;
+    (lo, hi)
+}
+
+/// Split a 128 bits number into two 64 bits limbs
+/// It will fail if the number is more than 128 bits
+fn split_into_64_bit_limbs(x: Field) -> (u64, u64) {
+    // Safety: the limbs are constrained below
+    let (x_lo, x_hi) = unsafe { split64(x) };
+    assert(x as Field == x_lo as Field + x_hi as Field * two_pow_64);
+    (x_lo, x_hi)
+}
+
+#[field(bn254)]
+pub fn wrapping_mul128(x: Field, y: Field) -> Field {
```

Comment:
This shouldn't need to be `pub` it looks like.

Reply:
Done